### PR TITLE
[Poetry]

### PR DESCRIPTION
**Pull request by shmuel44**
In Progress

## Related Issues
fixes: [link](https://jira-hq.paloaltonetworks.local/browse/CIAC-3411) to the issue

## Description
 * Changed package management from pip to Poetry for better dependency handling and packaging.
* Replaced `tox` with `Poetry` to run `pytest`.
* Changed the publish process to PyPI from using `twine` to `Poetry`.
* Update the README.md to reflect the new changes.
* Moved the handling of `Poetry` publish to `CircleCI`.

## Must have
- [ ] Unit Test or Example Code
- [ ] Changelog entry

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -2,41 +2,107 @@
...
+    docker:
+      - image: cimg/python:3.10-node
+    steps:
+      - checkout
+      - <<: *poetry_build
+      - run:
+          name: Publish Package PyPI
+          command: |
+            poetry --version 
+            # poetry publish -u __token__ -p ${}
```
Comment (dorschw) [inline .circleci/config.yml:81]: what's `${}` for?

File: .circleci/config.yml [hunk 2]
```diff
@@ -2,41 +2,107 @@
...
+              pythonversion: ["3.8", "3.9", "3.10"]   
+          name: run-unit-tests-<< matrix.pythonversion >>
+      - verify
+      - publish-to-test-pypi:
+          requires:
+            - run-unit-tests
+            - verify
+          filters:
+            branches:
+              only: sk_change_to_-poetry  # master
```
Comment (dorschw) [inline .circleci/config.yml:98]: master

File: .circleci/config.yml [hunk 3]
```diff
@@ -2,41 +2,107 @@
...
+      - publish-to-pypi:
+          requires:
+            - run-unit-tests
+            - verify
           filters:
             tags:
-              only: /^v\d+\.\d+\.\d+$/
+              only: /^test_sk_2+$/ #/^v\d+\.\d+\.\d+$/ # only run on tags [e.g. v0.0.0]
```
Comment (dorschw) [inline .circleci/config.yml:105]: undo

File: README.md [hunk 4]
```diff
@@ -142,47 +149,42 @@ When you open a new pull request, a bot will evaluate whether you have signed th
...
-Then setup dev virtual envs for python 3 (will also install all necessary requirements):
+You have now setup the your `demisto-py` dev environment!
 
-```bash
-tox --devenv py310
-```
+To activate it simply run: `poetry shell`.
 
-Switch to python 3 env by running:
+To deactivate the virtual environment and return simply run: `exit`.
```
Comment (dorschw) [inline README.md:171]: not `deactivate`?

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        'License :: OSI Approved :: Apache Software License',
+        'Programming Language :: Python',    
+        'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
+        'Programming Language :: Python :: 3.10',
+        'Programming Language :: Python :: Implementation :: CPython'
+    ]
+
+[tool.poetry.dependencies]
+python = ">=3.8.1"
```
Comment (dorschw) [inline pyproject.toml:23]: why not `3.8`?

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+[tool.poetry]
+name = "demisto-py"
+version = "3.3.1beta2"
```
Comment (dorschw) [inline pyproject.toml:3]: change before merging

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+description = "\"A Python library for the Demisto SDK\""
```
Comment (dorschw) [inline pyproject.toml:4]: not exactly

Comment (shmuel44) [reply pyproject.toml:4]: copy-paste from `setup.py`
` description="A Python library for the Demisto API"`

File: pyproject.toml (hunk 6 above)
Comment (shmuel44) [reply pyproject.toml:3]: Okay

File: pyproject.toml (hunk 5 above)
Comment (shmuel44) [reply pyproject.toml:23]: i changed it to 
`python = ">=3.8,<3.11"`
like in the `demisto-sdk`

File: README.md (hunk 4 above)
Comment (shmuel44) [reply README.md:171]: `To deactivate the virtual environment and exit this new shell type exit.` 
`To deactivate the virtual environment without leaving the shell use deactivate.`

[poetry-docs](https://python-poetry.org/docs/basic-usage/#activating-the-virtual-environment)

File: .circleci/config.yml [hunk 8]
```diff
@@ -2,41 +2,107 @@
...
+      - image: cimg/python:3.10-node
+    steps:
+      - checkout
+      - <<: *poetry_build
+      - run:
+          name: Publish Package Test PyPI
+          command: |
+            poetry --version
+            # poetry config repositories.testpypi https://test.pypi.org/legacy/
+            # poetry publish -u __token__ -p ${} -r testpypi
```
[the remark this answers is not in the document]
Comment (shmuel44) [reply .circleci/config.yml:68]: undo now

File: .circleci/config.yml (hunk 1 above)
Comment (shmuel44) [reply .circleci/config.yml:81]: undo now

File: .circleci/config.yml (hunk 2 above)
Comment (shmuel44) [reply .circleci/config.yml:98]: done

File: .circleci/config.yml [hunk 9]
```diff
@@ -2,41 +2,107 @@
...
+      - checkout
+      - run:
+          name: Verify
+          command: |
+            ./verify.sh
+  
+  publish-to-test-pypi:
+    docker:
+        # python 3.10 is the latest supported version
+      - image: cimg/python:3.10-node
```
[the remark this answers is not in the document]
Comment (shmuel44) [reply .circleci/config.yml:59]: del this, dine

File: .circleci/config.yml (hunk 3 above)
Comment (shmuel44) [reply .circleci/config.yml:105]: done

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        'Programming Language :: Python :: 3.10',
+        'Programming Language :: Python :: Implementation :: CPython'
+    ]
+
+[tool.poetry.dependencies]
+python = ">=3.8,<3.11"
+certifi = ">=2017.4.17"
+six = ">=1.10"
+python-dateutil = ">=2.5.3"
+setuptools = ">=21.0.0"
```
Comment (shmuel44) [inline pyproject.toml:27]: @dorschw do we need to delete the `setuptools`?

File: CHANGELOG.md [hunk 11]
```diff
@@ -3,6 +3,14 @@
 [PyPI History][1]
 [1]: https://pypi.org/project/demisto-py/#history
 
+## Unreleased
+
+* Changed package management from pip to Poetry for better dependency handling and packaging.
+* Replaced `tox` with `Poetry` to run `pytest`.
+* Changed the publish process to PyPI from using `twine` to `Poetry`.
+* Update the README.md to reflect the new changes.
+* Moved the handling of `Poetry` publish to `CircleCI`.
```
Comment (dorschw) [inline CHANGELOG.md:12]:
```suggestion
```
This is mostly for changes that affect the python library, not infrastructure ones

File: README.md [hunk 12]
```diff
@@ -142,47 +149,42 @@ When you open a new pull request, a bot will evaluate whether you have signed th
...
 
-```bash
-python -m pytest -v
+Simply use `poetry` to run your tests on a relevant folder. For example:
+```
+poetry run python -m pytest tests
 ```
 
-Additionally, our build uses tox to run on multiple envs. To use tox to run on all supported environments (py38, py39, py310), run:
+If you want to run with a specific python version (see the [demisto-py PyPi page](https://pypi.org/project/demisto-py/) for the current supported versions), use [poetry environments](https://python-poetry.org/docs/managing-environments/) to switch to a different env with a different python version.
```
Comment (dorschw) [inline README.md:184]:
```suggestion
To run with a specific python version (see the [demisto-py PyPi page](https://pypi.org/project/demisto-py/) for the current supported versions), use [poetry environments](https://python-poetry.org/docs/managing-environments/) to switch to a different env with a different python version.
```

**Review by dorschw — DISMISSED** [review]

File: pyproject.toml (hunk 10 above)
Comment (dorschw) [reply pyproject.toml:27]: yes

Comment (shmuel44) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 7 above)
Comment (shmuel44) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 6 above)
Comment (shmuel44) [reply pyproject.toml]: Done
3.2.12

**Review by dorschw — APPROVED** [review]

**Pull request merged**